### ISSUE TITLE
`allow_implicit_indents = True` not working for case when

### ISSUE THREAD

**Issue by kieronellis** [sqlfluff/sqlfluff#4579]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff fix is moving the first `when` statement to a new line, despite setting `allow_implicit_indents = True`.

### Expected Behaviour

As the documentation for [implicit indents](https://docs.sqlfluff.com/en/stable/layout.html#implicit-indents) states:
```
An implicit indent is exactly like a normal indent, but doesn’t have to be actually taken to influence the indentation of lines after it - it just needs to be left un-closed before the end of the line. These are normally available in clauses which take the form of KEYWORD <expression>, like WHERE clauses or CASE expressions.

-- This WHERE clause here takes advantage of an implicit indent.
SELECT *
FROM my_table
WHERE condition_a
   AND condition_b;

-- With implicit indents disabled (which is currently the
-- default), the above formulation is not allowed, and instead
-- there should be a newline immediately after `WHERE` (which
-- is the location of the _implicit_ indent).
SELECT *
FROM my_table
WHERE
   condition_a
   AND condition_b;
```

Therefore the following SQL should not be formatted:
```
select
  case old_col when 'aaaaaaaaaaaaaaaaaaaaaaaaa-a' then 'aaaaaaaaaaaaaaaaaaaaaaaaa1'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-b' then 'aaaaaaaaaaaaaaaaaaaaaaaaa2'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-c' then 'aaaaaaaaaaaaaaaaaaaaaaaaa3'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-d' then 'aaaaaaaaaaaaaaaaaaaaaaaaa4'
  end as new_col
from my_table
```

### Observed Behaviour

Instead the above SQL gets formatted to 
```
select
  case old_col
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-a' then 'aaaaaaaaaaaaaaaaaaaaaaaaa1'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-b' then 'aaaaaaaaaaaaaaaaaaaaaaaaa2'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-c' then 'aaaaaaaaaaaaaaaaaaaaaaaaa3'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-d' then 'aaaaaaaaaaaaaaaaaaaaaaaaa4'
  end as new_col
from my_table
```

Moving the first when statement to a new line and ignoring the config option `allow_implicit_indents = True`.

### How to reproduce

Create a .sql file with the contents:
```
select
  case old_col when 'aaaaaaaaaaaaaaaaaaaaaaaaa-a' then 'aaaaaaaaaaaaaaaaaaaaaaaaa1'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-b' then 'aaaaaaaaaaaaaaaaaaaaaaaaa2'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-c' then 'aaaaaaaaaaaaaaaaaaaaaaaaa3'
    when 'aaaaaaaaaaaaaaaaaaaaaaaaa-d' then 'aaaaaaaaaaaaaaaaaaaaaaaaa4'
  end as new_col
from my_table
```

and run `sqlfluff fix -f my_file.sql` with the below configuration.

### Dialect

bigquery

### Version

2.0.1

### Configuration

[sqlfluff]
templater = dbt
dialect = bigquery
large_file_skip_byte_limit = 30000
processes = -1
max_line_length = 120
runaway_limit = 100

[sqlfluff:indentation]
tab_space_size = 2
allow_implicit_indents = True  # Default = False.

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
# Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:layout.long_lines]
# Line length
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:layout:type:comma]
line_position = leading

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
# Data Types
extended_capitalisation_policy = lower

[sqlfluff:rules:convention.quoted_literals]
# Consistent usage of preferred quotes for quoted literals
preferred_quoted_literal_style = single_quotes

[sqlfluff:templater:dbt]
apply_dbt_builtins = True

[sqlfluff:templater:dbt:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Currently the indent before the `WHEN` is not configured to be an implicit indent, which I think is as designed. Perhaps it is worth making the documentation more explicit here that not all indents are configured to take advantage of implicit indents so that this doesn't cause more confusion.

@kieronellis - are you aware of any published style guides advocating for the layout you describe? It's not a format I've seen before.

**Comment by kieronellis** [comment]
Thanks for your replies.

@alanmcruickshank If it is not configured to be an implicit indent then what does the [documentation](https://docs.sqlfluff.com/en/stable/layout.html#implicit-indents) refer to here when it mentions `CASE expressions`?

```...These are normally available in clauses which take the form of KEYWORD <expression>, like WHERE clauses or CASE expressions.```

**Comment by alanmcruickshank** [comment]
Ah, that's a good question. I'll give some examples:

```sql
-- This is the "vanilla" layout. No indents at all:
SELECT
    CASE
        WHEN a AND b THEN c + c
        WHEN d AND e THEN f + f
        ELSE g + g
    END AS h
FROM my_table;

-- This is the layout with line breaks, but _without_ implicit indents
SELECT
    CASE
        WHEN
            a
            AND b
            THEN
                c
                + c
        WHEN
            d
            AND e
            THEN
                f
                + f
        ELSE
            g
            + g
    END AS h
FROM my_table

-- Enabling implicit indents allows this alternative layout:

SELECT
    CASE
        WHEN a
            AND b
            THEN c
                + c
        WHEN d
            AND e
            THEN f
                + f
        ELSE g
            + g
    END AS h
FROM my_table
```

You'll see that the implicit indents are more about the indentation of the expressions _within_ the `WHEN`, `THEN` and `ELSE` clauses rather than the placement of the `WHEN` clause itself. Is that clearer than the phrase in the docs?

I can see that the docs probably need to be clearer on this front.

**Comment by alanmcruickshank** [comment]
That being said - #4587 is a proposed solution to this.

**Comment by alanmcruickshank** [comment]
I think this was resolved by #4598 so I'm going to close this issue.
